_posts/2018-05-31-regular-types.md

```diff
@@ -0,0 +1,874 @@
+describe the syntax and semantics of built-in types in a fashion that allows
+user-defined types to behave sensibly.
+
+"The C++ programming language allows the use of built-in type operator syntax
+for user-defined types. This allows us, as programmers, to make our user-defined
+types look like built-in types. Since we wish to extend semantics as well as
+syntax from built-in types to user types, we introduce the idea of a Regular
+type, which matches the built-in type semantics, thereby making our user-defined
+types behave like built-in types as well."
+
+A vastly-simplified summary, that has become popular in C++ design in later
+years, is "do as the ints do." Generally speaking, for a snippet of generic code
+operating on some type `T`, if it works properly on ints and also works properly
+for your new type (similarly bug-free/comprehensible/etc.), you've designed a
+reasonable type.
+
+## Definitions of Regular
+
+In fact, this evaluation of generic code is inherent in the (somewhat flexible)
+definitions for Regular; Stepanov defines certain properties for Regular, but
+different formulations may include a slightly different set of requirements. In
+particular, proposal [P0898](http://wg21.link/p0898) aims to define Concepts for
+use in the C++ standard library; the requirements for Regular as presented in
+P0898 are somewhat reduced from Stepanov's definition (ordering is not
+required). The crux of this difference seems to be that Stepanov's definition
+stems from the requirements of a type with respect to the full set of C++98-era
+standard algorithms, while P0898 focuses primarily on the requirements to be
+used in a more modern library (specifically, the Ranges library as specified in
+Range v3 and/or the Ranges TS).
+
+<table>
+<thead>
+<tr>
+<th><strong>Stepanov's Regular Requirements</strong></th>
+<th><strong>P0898 Concept Requirements for Regular</strong></th>
+</tr>
+</thead>
+<tbody>
+<tr>
+<td>Default constructor</td>
+<td>DefaultConstructible</td>
+</tr>
+<tr>
+<td>Copy constructor</td>
+<td>CopyConstructible</td>
+</tr>
+<tr>
+<td>Destructor</td>
+<td>Destructible</td>
+</tr>
+<tr>
+<td>Movable?</td>
+<td>Movable</td>
+</tr>
+<tr>
+<td></td>
+<td>Swappable</td>
+</tr>
+<tr>
+<td>Assignment</td>
+<td>Assignable</td>
+</tr>
+<tr>
+<td>Equality</td>
+<td>EqualityComparable</td>
+</tr>
+<tr>
+<td>Inequality</td>
+<td>EqualityComparable</td>
+</tr>
+<tr>
+<td>Total Ordering</td>
+<td></td>
+</tr>
+</tbody>
+</table>
+
+(It’s generally assumed that Move construction would be included in Stepanov’s
+formulation, although move semantics were not present in C++ of that era. This
+is true even though `int` doesn’t have a move constructor: it is still move
+constructible (can be constructed from a temporary), and this is even the proper
+design for the type. Move+copy should be considered an overload set for
+optimization purposes. More in [TotW 148](http://abseil.io/tips/148))
```

Comment:
Missing a "." in the parenthetical.

Reply:
Done. Removed first parenthetheses, btw, as that resulted in nested parens